server_client/src/renderer/LeftPanel.tsx

```diff
@@ -218,30 +234,66 @@ export const LeftPanel = (props: { pages: Page[] }) => {
   const [stretchButtons, setStretchButtons] = useState(false);
   const [colWidth, setColWidth] = useState(150);
   const [rowHeight, setRowHeight] = useState(150);
+  const [pageHeight, setPageHeight] = useState(0);
+  const [minPageHeight, setMinPageHeight] = useState(0);
+  const [minPageWidth, setMinPageWidth] = useState(0);
+  const [pageWidth, setPageWidth] = useState(0);
+  const [pageId, setPageId] = useState(pages[0].id);
   const [page, setPage] = useState(pages[0]);
 
   useEffect(() => {
-    // If the pages data gets modified externally (i.e the advanced editor), we
-    //   need to refresh the data here
-    setPage(pages[0]);
-  }, [pages]);
+    const determineMinWidthHeight = (p: Page) => {
+      let highestX = 0;
+      let highestY = 0;
+
+      for (let y = 0; y < p.height - 1; y++) {
+        for (let x = 0; x < p.width - 1; x++) {
+          const button = p.buttons.find((b) => b.x === x && b.y === y);
+          if (button) {
+            if (y > highestY) {
+              highestY = y;
+            }
+            if (x > highestX) {
+              highestX = x;
+            }
+          }
+        }
+      }
+      setMinPageHeight(highestY + 1);
+      setMinPageWidth(highestX + 1);
+    };
+    // If the pages data gets modified externally or we change the current page ID
+    //    refresh the data and show the page
+    const newPage = pages.find((p) => p.id === pageId);
+    if (newPage) {
+      setPage(newPage);
+      setPageHeight(newPage.height);
+      setPageWidth(newPage.width);
+      determineMinWidthHeight(newPage);
+    } else {
+      setPage(pages[0]);
+      setPageHeight(pages[0].height);
+      setPageWidth(pages[0].width);
+      determineMinWidthHeight(pages[0]);
+    }
+  }, [pages, pageId]);
 
   useEffect(() => {
     const setSize = () => {
       // I truly hate doing this this way, but every way I've seen of doing this in
-      //    CSS looks equally bad in my mind, and this is more readable.
+      //    CSS looks equally bad in my mind, and this is way more readable.
       let w = window.innerWidth;
       let h = window.innerHeight;
 
       w -= 300; // Sidebar
       w -= 8; // padding
       h -= 8; // padding
       h -= 50; // bottom bar
-      h -= 20; // margin of error
-      w -= 20; // margin of error
+      h -= 50; // margin of error
+      w -= 50; // margin of error
```

Comment:
Makes it play a bit nicer with more extreme values, not really noticeable with smaller numbers